File: src/17장/17-3.py
k, n = map(int, input().split())
lanson = []
for i in range(k): lanson.append(int(input()))
low, high = 0, 10000000000
answer=0
while low <= high:
    mid = (low + high) // 2
    num = 0
    for len in lanson: 
      num += len//mid 
    if num >= n: 
      low = mid + 1
      if mid> answer :
        answer=mid
    else: high = mid - 1
print(answer)

File: src/17장/17-3.java
import java.util.*;
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
		Scanner sc = new Scanner(System.in);
        int K = sc.nextInt();
        long N = sc.nextLong();
        long[] lanson = new long[K];
        long answer = 0;
        for (int i = 0; i < K; i++) {
            lanson[i] = sc.nextLong();
        }
        long low = 0; 	  
        long high = (long)2147483647;
        while (low <= high) {
            long mid = (low + high) / 2;
            long num = 0;
            for (int i = 0; i < K; i++) { 
                num += (lanson[i] / mid);
            }
            if (num >= N) { 
                low = mid + 1;
				answer=mid;
            } else {
                high = mid - 1;
            }
        }
        System.out.println(answer);
	}
}